dpcpp/matrix/csr_kernels.dp.cpp

```diff
@@ -71,11 +78,1820 @@ namespace dpcpp {
+                                 orig_col_idxs, diag, item_ct1);
+            });
+    });
+}
+
+
+}  // namespace kernel
+
+
+namespace {
+
+
+template <typename ValueType>
+void conjugate_kernel(size_type num_nonzeros, ValueType *__restrict__ val,
+                      sycl::nd_item<3> item_ct1)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+
+    if (tidx < num_nonzeros) {
+        val[tidx] = conj(val[tidx]);
+    }
+}
+
+template <typename ValueType>
+void conjugate_kernel(dim3 grid, dim3 block, size_type dynamic_shared_memory,
+                      sycl::queue *queue, size_type num_nonzeros,
+                      ValueType *val)
+{
+    queue->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(sycl_nd_range(grid, block),
+                         [=](sycl::nd_item<3> item_ct1) {
+                             conjugate_kernel(num_nonzeros, val, item_ct1);
+                         });
+    });
+}
+
+
+}  //  namespace
+
+
+template <typename IndexType>
+void inv_permutation_kernel(size_type size,
+                            const IndexType *__restrict__ permutation,
+                            IndexType *__restrict__ inv_permutation,
+                            sycl::nd_item<3> item_ct1)
+{
+    auto tid = thread::get_thread_id_flat(item_ct1);
+    if (tid >= size) {
+        return;
+    }
+    inv_permutation[permutation[tid]] = tid;
+}
+
+template <typename IndexType>
+void inv_permutation_kernel(dim3 grid, dim3 block,
+                            size_type dynamic_shared_memory, sycl::queue *queue,
+                            size_type size, const IndexType *permutation,
+                            IndexType *inv_permutation)
+{
+    queue->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(sycl_nd_range(grid, block),
+                         [=](sycl::nd_item<3> item_ct1) {
+                             inv_permutation_kernel(size, permutation,
+                                                    inv_permutation, item_ct1);
+                         });
+    });
+}
+
+
+template <typename ValueType, typename IndexType>
+void col_permute_kernel(size_type num_rows, size_type num_nonzeros,
+                        const IndexType *__restrict__ permutation,
+                        const IndexType *__restrict__ in_row_ptrs,
+                        const IndexType *__restrict__ in_cols,
+                        const ValueType *__restrict__ in_vals,
+                        IndexType *__restrict__ out_row_ptrs,
+                        IndexType *__restrict__ out_cols,
+                        ValueType *__restrict__ out_vals,
+                        sycl::nd_item<3> item_ct1)
+{
+    auto tid = thread::get_thread_id_flat(item_ct1);
+    if (tid < num_nonzeros) {
+        out_cols[tid] = permutation[in_cols[tid]];
+        out_vals[tid] = in_vals[tid];
+    }
+    if (tid <= num_rows) {
+        out_row_ptrs[tid] = in_row_ptrs[tid];
+    }
+}
```

Comment:
when I write it in place, where can I write the __restrict__?
they are captured by lambda

Reply:
Is it not possible to add `__restrict__` to the kernel function? Is that not forwarded to the in-place lambda? I guess that would need testing, but I would much rather have that than the current additional lambda, which causes a lot of boiler-plate code for only the `__restrict__`. If that is not possible, we can keep the current code.

Reply:
another issue comes to my mind.
if I put the function inside  the lambda directly, I seems to need to take similar amount lines for it. (if it is mainly from #lines concern)
I need to take the plain pointer first and then can put it in lambda function.
Otherwise, the lambda will try to capture the whole class not just the pointer, which is not allowed.
It will be like following
```
const IndexType *__restrict__ permutation = permutation.get_values();
// .... for each pointer.
and then
cgh.parallel_for(, {...});
```
these may be suitable for common (generic) kernel

Reply:
I would actually prefer this solution (for smaller kernels). Not just for reducing the number of lines, but also to make reading the kernel easier.

Reply:
after revisiting it, if using DEFAULT_HOST, I think they are similar.
when you search col_permute_kernel, only one place has the code.
but this should be moved to common part